Add search view

fanwgwg opened the pull request:
Add a search view that will display structured results while typing, based on substring search.

![capture](https://cloud.githubusercontent.com/assets/14089267/23738159/627c9cc4-04d2-11e7-8f4f-f0beba78103a.gif)

Benglin commented on src/LibraryUtilities.ts:
```diff
@@ -249,3 +255,32 @@ export function buildLibraryItemsFromLayoutSpecs(loadedTypes: any, layoutSpecs:
     return convertToLibraryTree(typeListNodes, layoutElements);
 }
+
+function setItemVisible(item: ItemData) {
```
Let's name this `showItemRecursive` instead, making the intents very clear.

Benglin commented on src/LibraryUtilities.ts:
```diff
@@ -249,3 +255,32 @@ export function buildLibraryItemsFromLayoutSpecs(loadedTypes: any, layoutSpecs:
+}
+
+export function search(text: string, item: ItemData) {
+    let index = item.text.toLowerCase().indexOf(text);
```
I think once `index` is set and it is not `-1`, then we can make the item visible and return, right? Otherwise the `for` loop below will overwrite its value.

Benglin replied:
Shouldn't the above `indexOf` (and also the `for` loop below) to be done only for items that are not `group` type? If we can directly skip these steps for `group` items then it'll save time.

Benglin commented on src/LibraryUtilities.ts:
```diff
@@ -249,3 +255,32 @@ export function buildLibraryItemsFromLayoutSpecs(loadedTypes: any, layoutSpecs:
+    }
+
+    if (index >= 0 && item.itemType !== "group") {
+        setItemVisible(item); // show all child items if text is found in parent item
```
Do update the comments to be like so:

```js
if (index >= 0 && item.itemType !== "group") {
    // Show all items recursively if a given text is found in the current 
    // (parent) item. Note that this does not apply to items of "group" type
    setItemVisible(item);
}
```

fanwgwg commented:
@Benglin I've just made some changes regarding your comment.

fanwgwg commented:
@Benglin Can merge this? Then I'll make another pull request for the list view

Benglin commented on src/LibraryUtilities.ts:
```diff
@@ -248,3 +254,39 @@ export function buildLibraryItemsFromLayoutSpecs(loadedTypes: any, layoutSpecs:
+
+export function search(text: string, item: ItemData) {
+    if (item.itemType !== "group") {
+        let index: number;
```
Can we explicitly give `index` a value of `-1` here instead of leaving it unassigned? Because its value will be unknown if `item.searchStrings` loop below does not execute due to empty list.

Benglin commented on src/components/LibraryItem.tsx:
```diff
@@ -54,7 +57,21 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
+        }
+        if (nextProps.data.expanded !== this.state.expanded) {
+            this.setState({ expanded: nextProps.data.expanded });
+        }
```
Can you describe why is this `componentWillReceiveProps` needed (and what's the difference between this and the `constructor` above)? If this is needed, can we not merge both the `setState` into one? That way React.js doesn't have to queue up two state changes since they are aiming at the same component.

Benglin replied:
I meant to suggest this:

```js
if ((nextProps.data.visible !== this.state.visible) || 
    (nextProps.data.expanded !== this.state.expanded)) {
        this.setState({
            visible: nextProps.data.visible,
            expanded: nextProps.data.expanded
        });
}
```

fanwgwg replied:
We will need this if we want to allow user to toggle the expansion in the structured search view. Since by default all the found items in search view have `expanded = true` stored in their `ItemData`, we also have a `expanded` state in the state of `LibraryItem`, this is needed to make sure that these two always the same. Otherwise there will be issues like which one should we check when rendering. However, if we just don't allow user to unexpand in search view, this will not be needed.

Benglin commented on src/components/LibraryItem.tsx:
```diff
@@ -73,31 +90,33 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
-        {
-            if (this.props.data.childItems && (this.props.data.childItems.length > 0)) {
+        // visible only nested elements when expanded.
+        if (this.state.expanded && this.state.visible && this.props.data.childItems.length > 0) {
```
`this.props.data.childItems` condition is removed, is it always non-undefined?

Also this may look cleaner:

```js
let shouldDisplay = this.state.expanded && this.state.visible;
if (shouldDisplay && (this.props.data.childItems.length > 0)) {
    // ...
}
```

fanwgwg replied:
Actuallt in the `ItemData` class, we initialize `childItems: ItemData[] = []`, so it will not be undefined.

Benglin commented on src/components/LibraryItem.tsx:
```diff
@@ -73,31 +90,33 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
+            );
+        } else {
+            return null;
+        }
```
Do we come all the way here (i.e. after computing all other stuff like `nestedElements`) just to `return null` when `this.state.visible === false`? Remember that will incur unnecessary calculations that can be avoided? If that's the case move `return null` all the way up (and remove other `if (this.state.visible)` related checks since it will be `true` once it passed the initial check.

Benglin commented on src/components/SearchView.tsx:
```diff
@@ -0,0 +1,81 @@
+import { search, ItemData } from "../LibraryUtilities";
+
+interface SearchViewProps {
+    onSearchChanged: any;
```
This should be renamed to `onSearchModeChanged`, and it would be nice if you can [give it a type](https://www.typescriptlang.org/docs/handbook/interfaces.html) like so:

```js
interface SearchModeChangedFunc {
    (inSearchMode: boolean): void;
}

interface SearchViewProps {
    onSearchModeChanged: SearchModeChangedFunc;
}
```

Benglin commented on src/components/SearchView.tsx:
```diff
@@ -0,0 +1,81 @@
+        let index = 0;
+        structuredItems = this.props.items.map((item: ItemData) => <LibraryItem key={index++} libraryView={this.props.libraryView} data={item} />);
+        return structuredItems;
+    }
```
Do clean it up this way:

```js
generateStructuredItems() {
    let index = 0;
    return this.props.items.map((item: ItemData) => 
        <LibraryItem key={index++} libraryView={this.props.libraryView} data={item} />);
}
```

Benglin commented on src/components/SearchView.tsx:
```diff
@@ -0,0 +1,81 @@
+            item.visible = visibility;
+            this.setItemVisibility(visibility, item.childItems);
+        });
+    }
```
Move `setItemVisibility` to utility class, I don't think `this` is required at all.

Benglin commented on src/components/SearchView.tsx:
```diff
@@ -0,0 +1,81 @@
+            item.expanded = expandability;
+            this.setItemExpandability(expandability, item.childItems);
+        });
+    }
```
I got the feeling that we are enabling/disabling the expandability here, I think a better term for `setItemExpandability` would `setItemExpandState` be nicer?

Benglin replied:
Move `setItemExpandability` to `LibraryUtilities.ts` and add test cases for it, I don't think `this` is required at all.

Benglin commented on src/components/SearchView.tsx:
```diff
@@ -0,0 +1,81 @@
+
+    performSearch(text: string) {
+        this.props.items.forEach(item => search(text, item));
+    }
```
Move `performSearch` to utility class and add tests for it, maybe name it `searchItemRecursive`.

Benglin commented on src/components/LibraryContainer.tsx:
```diff
@@ -13,7 +14,9 @@ export interface LibraryContainerProps {
 
-export interface LibraryContainerStates { }
+export interface LibraryContainerStates {
+    inSearch: boolean
```
I think `inSearchMode` can be a better name.

Benglin commented on src/LibraryUtilities.ts:
```diff
@@ -248,3 +254,39 @@ export function buildLibraryItemsFromLayoutSpecs(loadedTypes: any, layoutSpecs:
+function showItemRecursive(item: ItemData) {
+    item.visible = true;
+    item.childItems.forEach((childItem) => showItemRecursive(childItem));
+}
```
Move `showItemRecursive` method to utility class, then add test cases for it.

Benglin commented on src/components/LibraryContainer.tsx:
```diff
@@ -22,18 +25,38 @@ export class LibraryContainer extends React.Component<LibraryContainerProps, Lib
 
+    onSearchChanged(inSearch: boolean) {
+        this.setState({ inSearch: inSearch });
+    }
```
Do rename this function to `onSearchModeChanged` too.

fanwgwg commented:
@Benglin Thanks a lot for the review! 
I've just updated the code accordingly, and added some testcases for those search-related functions in LibraryUtilities

Benglin commented on src/LibraryUtilities.ts:
```diff
@@ -245,3 +251,54 @@ export function buildLibraryItemsFromLayoutSpecs(loadedTypes: any, layoutSpecs:
+    for(let childItem of item.childItems) {
+        showItemRecursive(childItem);
+    }
+}
```
Hi @fanwgwg, thanks for doing this. In a subsequent pull request you can merge `resetItemData` and `showItemRecursive` into one that looks like this:

```js
export function setItemStateRecursive(items: ItemData[], visible: boolean, expanded: boolean) {
    // Makes use of visible and expanded parameters
}
```

Benglin commented on src/components/LibraryContainer.tsx:
```diff
@@ -22,18 +25,37 @@ export class LibraryContainer extends React.Component<LibraryContainerProps, Lib
     render() {
 
         try {
             let index = 0;
```
You can do this later, but let's move `index` into the `else` clause below before the `map`.

Benglin commented on src/components/LibraryItem.tsx:
```diff
@@ -54,10 +60,23 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
+    componentWillReceiveProps(nextProps: LibraryItemProps) {
+        if (nextProps.data.expanded !== this.state.expanded) {
+            this.setState({ expanded: nextProps.data.expanded });
+        }
```
Please add some comments to describe why is this needed in a subsequent PR.

Benglin commented on src/components/SearchView.tsx:
```diff
@@ -0,0 +1,66 @@
+        return (
+            <div className="searchView">
+                <div className="searchBar">
+                    <input id="searchInput" type="search" placeholder="Search..." onChange={this.onTextChange.bind(this)}></input>
```
Hi @fanwgwg, can you make a task for this? We want `onChange` to not trigger a search so quickly, it should start a timer of maybe `300ms` and if that time expired then we will trigger the search. Otherwise if a new character is being pressed, then the timer will restart to be `300ms`. Thanks!

fanwgwg replied:
ok sure! I've just made some changes according to the preivous comments.

Benglin commented:
Thanks again for being so innovative on the search view, we didn't think of that being possible before so good job there! Please follow up with other comments in this PR later.
